text/0000-multi-api-lifecycle-descriptor.md

```diff
@@ -0,0 +1,172 @@
+**When** the `api` field in a `buildpack.toml` file is set to an API version in the deprecated buildpack API range the lifecycle shall:
+ - **If** `CNB_BUILDPACK_DEPRECATION_MODE` is unset, **Then** print a warning and continue
+ - **If** `CNB_BUILDPACK_DEPRECATION_MODE=warn`, **Then** print a warning and continue
+ - **If** `CNB_BUILDPACK_DEPRECATION_MODE=error`, **Then** fail
+ - **If** `CNB_BUILDPACK_DEPRECATION_MODE=silent`, **Then** continue w/o warning
+
+### Supported APIs
+Only API versions defined in a spec release can be in the supported range.
+
+Supported APIs will behave as expected. 
+
+### Experimental APIs
+Experimental API versions may or may not correspond to a spec release.
+
+The behavior of experimental APIs may change between lifecycle versions without notice.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Platforms
+Platforms should use the contents of `lifecycle.toml` to set `io.buildpacks.lifecycle.version`, and `io.buildpacks.lifecycle.apis` labels when creating builders.
+
+The lifecycle will set `io.buildpacks.lifecycle.version` and `io.buildpacks.lifecycle.apis` on all future released lifecycle images.
+
+Platforms may use the contents of the `io.buildpacks.lifecycle.apis` label when deciding which `CNB_PLATFORM_API` to use, or validating whether a buildpack is compatible with a given builder image.
+
+### `pack`
+`pack inspect-builder` should display deprecated, supported, and experimental API information when it is available.
+
+## Backwards Compatibility
+
+Currently lifecycle API and verison information is contained in the `io.buildpacks.builder.metadata` label. The following is an abridged example label
+```
+{
+  "buildpacks": [],
+  "created-by": {}
+  "description": "",
+  "lifecycle": {
+    "version": "<lifecycle version>",
+    "api": {
+      "buildpack": "<buildpack API version>",
+      "platform": "<platform API version>"
+    }
+  },
+  "stack": {},
+}
+```
+For backwards compatibility `pack create-builder` should continue to populate the above label, setting `lifecycle.api.buildpack`, and `lifecycle.api.platform` to the corresponding supported `min`.
+
+This will allow platforms the read the metadata in it's existing format to update gracefully. Eventually this compatibility layer will be removed.
+
+The lifecycle descriptor file itself will not preserve backwards compatibility. Most platforms consume builders rather than directly consuming lifecycle artifacts, and therefore the builder label is the important compatibility concern. Those platforms that do consume lifecycle artifacts control which versions of the lifecycle they bring in and can update their behavior to parse the new format for newer lifecycles.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Understanding buildpack/platform API support will require more documentation and education.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- We could have `deprected` and `supported` lists instead of using `min` and `max` to define ranges.
```

Comment:
Just thinking out loud, but I assume there's never a point where we would deprecate a API in the middle of a range b/c we messed up badly.

Reply:
Yeah, I thought about this and I feel fairly confident we would never want to deprecate a API in the middle of a range. I think that would cause a major headache for end users and I can't foresee us wanting to anyways.

Reply:
I agree, I just wanted to throw the idea out there.